Java : add request forgery query

porcupineyhairs opened the pull request:
This query adds support for detecting SSRF in Java.

porcupineyhairs commented:
I just noticed #3452 tackles a similar issue. I will give a brief of how my PR differs from the latter.

- That one adds support for only `openConnection` calls. I add support for multiple other calls from the `java.net.http` and `apache http client` libraries. 

- I have also included library as well as query tests for the same. 

- I haven't included any barriers for the flow while #3452 considers addition with lhs as `://` as a blocking step.

- #3452 is asking to be merged as an experimental one while I would like this to be included as stable.

intrigus-lgtm commented on java/ql/src/semmle/code/java/frameworks/javase/URL.qll:
```diff
@@ -0,0 +1,56 @@
+  Expr protocolArg() {
+    // In all cases except where the first parameter is a URL, the argument
+    // containing the protocol is the first one, otherwise it is the second.
+    if this.getConstructor().getParameter(0).getType().getName() = "URL"
```
Can you not do `if this.getConstructor().getParameter(0).getType() instanceof TypeUrl` instead?

porcupineyhairs replied:
That's not my code. I refactor it from `java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql`. 
I will change it if @aschackmull is fine with that change though.

intrigus-lgtm commented on java/ql/test/library-tests/frameworks/javase/uri.java:
```diff
@@ -0,0 +1,60 @@
+import java.net.URLStreamHandler;
+import java.net.Proxy.Type;
+
+public class uri {
```
As far as I know, java classes are CamelCase, i.e. `Uri` instead of `uri`.

pwntester commented:
Can you please open a corresponding [securitylab issue](https://github.com/github/securitylab/issues) for tracking?

JLLeitschuh commented on java/ql/src/Security/CWE/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,70 @@
+  // when the builder is tainted
+  exists(MethodAccess m, RefType t |
+    m.getMethod().getDeclaringType() = t and
+    t.hasQualifiedName("org.springframework.http", "RequestEntity<>$BodyBuilder") and
```
Would be nice to turn this into a type instead of having it inline? Maybe?

porcupineyhairs replied:
Not doing that. I don't see any reason why someone would want to reuse this. Any how, if this is required in the future, this can be refactored then.

JLLeitschuh commented on java/ql/src/semmle/code/java/frameworks/javase/URI.qll:
```diff
@@ -0,0 +1,53 @@
+class URIConstructor extends ClassInstanceExpr, URICreation {
+  URIConstructor() { this.getConstructor().getDeclaringType().getQualifiedName() = "java.net.URI" }
+
+  override Expr hostArg() {
+    // URI​(String str)
+    if this.getNumArgument() = 1
+    then result = this.getArgument(0)
+    else
+      // URI​(String scheme, String ssp, String fragment) 	
+      if this.getNumArgument() = 3
+      then result = this.getArgument(1)
+      else
+        // URI​(String scheme, String userInfo, String host, int port, String path, String query,
+        //    String fragment) 	
+        if this.getNumArgument() = 7
+        then result = this.getArgument(2)
+        else
+          // URI​(String scheme, String host, String path, String fragment) 	
+          if this.getNumArgument() = 4
+          then result = this.getArgument(1)
+          else
+            // URI​(String scheme, String authority, String path, String query, String fragment)
+            if this.getNumArgument() = 5
+            then result = this.getArgument(1)
+            else none()
```
Could this be re-written as a type hierarchy instead? It may not be possible.

porcupineyhairs replied:
I don't think a type heirarchy would be a good choice here. I understand that this is a bad coding practice but I have no clue how to eliminate this from my code. @aschackmull Any thoughts on this?

intrigus-lgtm replied:
Couldn't something like this work?
```ql
result = this.getArgument(0) and this.getNumArgument() = 1
or
result = this.getArgument(1) and this.getNumArgument() = 3
or
result = this.getArgument(2) and this.getNumArgument() = 7
or
result = this.getArgument(1) and this.getNumArgument() = 4
or
result = this.getArgument(1) and this.getNumArgument() = 5
```
Same for `URL` below.

porcupineyhairs replied:
This may work for URI but not for URL as there are multiple functions with the same no of args but different signatures. This looks much better though. I will change my code to something similar to what you suggest.

JLLeitschuh replied:
> You can also shorten this further by using set literals:

🤯 New syntax! Very cool!

JLLeitschuh commented on java/ql/src/Security/CWE/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,70 @@
+    }
+  }
+
+  class RequestForgeryLocalConfiguration extends TaintTracking::Configuration {
```
In production code this probably happens quite a bit. In prod apps, you may have a configurable hostname that is read from a config file. I feel like this would most often cause false positives.

If you're already local on the machine, SSRF may be the least of your problems.

porcupineyhairs replied:
I just added this as all other queries add this.

JLLeitschuh commented on java/ql/src/Security/CWE/CWE-918/RequestForgery.java:
```diff
@@ -0,0 +1,20 @@
+		client.send(r, null);
+
+		// GOOD: the request parameter is validated against a known fixed string
+		if (VALID_URI.equals(request.getParameter("uri"))) {
```
I feel like I'm missing something. Where in the query do you have that logic like this guards against SSRF?

porcupineyhairs replied:
I don't have them as of now. it a false positive which we can tackle later if there are too many instances of this type. 
The snippet here is to be included in the qhelp as a correct way of preventing SSRF that's it.

porcupineyhairs commented:
I have added a few stubs which I had missed earlier and rebased this to the latest master. This is now ready to be reviewed and merged.

porcupineyhairs commented:
@aschackmull I have rebased to the latest main. This is now ready for review.

JLLeitschuh commented on java/ql/src/experimental/CWE-918/RequestForgery.java:
```diff
@@ -0,0 +1,20 @@
+	throws ServletException, IOException {
+		URI uri = new URI(request.getParameter("uri"));
+		// BAD: a request parameter is incorporated without validation into a Http request
+        HttpRequest r = HttpRequest.newBuilder(uri).build();
```
Formatting weirdness herre.

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringHttp.qll:
```diff
@@ -38,3 +39,21 @@ class SpringResponseEntityBodyBuilder extends Interface {
+// }
+/** Models `org.springframework.http.RequestEntity`s instantiation expressions. */
+class SpringRequestEntityInstanceExpr extends ClassInstanceExpr {
+  int numArgs;
```
Is this used anywhere?

porcupineyhairs replied:
https://github.com/github/codeql/blob/a210fbbe69909e14c90f09ef77eb8f157e05ee4e/java/ql/src/experimental/CWE-918/RequestForgeryCustomizations.qll#L242

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/URI.qll:
```diff
@@ -0,0 +1,43 @@
+}
+
+/* An URI constructor expression */
+class URIConstructor extends ClassInstanceExpr, URICreation {
```
Use `Uri` not `URI` everywhere (similarly other all-caps words in camel-case names)

smowton commented on java/ql/src/semmle/code/java/frameworks/Networking.qll:
```diff
@@ -19,6 +19,11 @@ class TypeUrl extends RefType {
 }
 
+/** The type `java.net.URI`. */
+class TypeUri extends RefType {
```
There are 3 other `hasQualifiedName("java.net", "URI")` uses in the codebase -- let's factor them (in a separate commit) while we're introducing this

porcupineyhairs replied:
Creating a separate PR to bring this change.

porcupineyhairs replied:
See the new PR here. https://github.com/github/codeql/pull/4600

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/Http.qll:
```diff
@@ -0,0 +1,20 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+
+/* A class representing Http Request Builder. */
```
Use backticks and the proper Java class name in doc comments (i.e., "representing `HttpRequest.Builder`")

smowton replied:
Also regular comments -> doc comments throughout

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringRestTemplate.qll:
```diff
@@ -0,0 +1,93 @@
+import semmle.code.java.frameworks.spring.SpringWebClient
+
+/**Models `RestTemplate` class's `doExecute` method */
+class RestTemplateDoExecute extends Method {
```
All code relating to these (as noted above I think we can factor a lot of it) should be moved to the existing file `SpringWebClient.qll`

smowton commented on java/ql/test/experimental/query-tests/security/CWE-918/RequestForgery.java:
```diff
@@ -0,0 +1,34 @@
+// import java.io.IOException;
+// import java.net.URI;
+// import java.net.http.HttpClient;
+// import java.net.http.HttpRequest;
```
What are these commented tests?

porcupineyhairs replied:
fixed.

smowton commented on java/ql/test/experimental/stubs/springframework-5.2.6/org/springframework/core/ParameterizedTypeReference.java:
```diff
@@ -0,0 +1,27 @@
+    public java.lang.reflect.Type getType() {
+        return null;
+    }
+
```
There are already stubs in the repo for springframework-5.2.3; can we use those instead of adding these?

porcupineyhairs replied:
fixed

smowton reviewed (commented):
Hello! Apologies for slowness on our part; I'm taking over reviewing and hope the process will be much quicker from here. Quite a few comments but nothing too time-consuming I hope.

porcupineyhairs commented on java/ql/src/experimental/CWE-918/RequestForgeryCustomizations.qll:
```diff
@@ -0,0 +1,246 @@
+   */
+  private class RestTemplateDoExecuteArg extends Sink {
+    RestTemplateDoExecuteArg() {
+      exists(MethodAccess ma | ma.getMethod() instanceof RestTemplateDoExecute |
```
Done! :+1:

porcupineyhairs commented:
@smowton Sorry for the delay in addressing the review. I was on a long vacation. I have made the changes and merged the latest master into the branch.

smowton commented on .gitignore:
```diff
@@ -13,6 +13,7 @@
 # Visual studio temporaries, except a file used by QL4VS
 .vs/*
 !.vs/VSWorkspaceSettings.json
+.vscode/settings.json
```
Remove this change

porcupineyhairs replied:
Done! Sorry about the multiple forced rebases. Git won't let me drop these files for some reason. Anyways, its all fixed now.

smowton commented on .vscode/settings.json:
```diff
@@ -1,3 +1,4 @@
 {
-    "omnisharp.autoStart": false
+    "omnisharp.autoStart": false,
+    "codeQL.runningQueries.autoSave": true
```
Remove this change

smowton commented on java/ql/src/experimental/CWE-918/RequestForgery.qll:
```diff
@@ -0,0 +1,69 @@
+    }
+  }
+
+  class RequestForgeryLocalConfiguration extends TaintTracking::Configuration {
+    RequestForgeryLocalConfiguration() { this = "Server Side Request Forgery" }
+
+    override predicate isSource(DataFlow::Node source) { source instanceof LocalUserInput }
+
+    override predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    override predicate isAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ) {
+      additionalStep(pred, succ)
+    }
+  }
```
No longer needed

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringWebClient.qll:
```diff
@@ -27,3 +27,116 @@ class SpringWebClient extends Interface {
+  override Argument getUrlArgument(MethodAccess ma) { result = ma.getArgument(0) }
+}
+
+/**Models `RestTemplate` class's `put` method */
```
```suggestion
/** Models `RestTemplate` class's `put` method */
```
Here and elsewhere

smowton reviewed (commented):
Just some minor comments, checking the evaluation status for this one

smowton commented:
Once your subsidiary PR is merged please *rebase* onto `main` (rather than merge it in)

smowton commented:
Looks like eval was done long ago, will flag this to codeql-java owners for review

aschackmull commented:
Could we get this rebased on main? The changes from https://github.com/github/codeql/pull/4600 still show up here.

porcupineyhairs commented:
rebased on the latest main!

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/URI.qll:
```diff
@@ -0,0 +1,43 @@
+   *  such as in `URI(`http://foo.com/mypath')`,
+   *  this returns the entire argument passed i.e. `http://foo.com/mypath'.
+   */
+
```
Linebreak probably means the doc comment won't bind to the method

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/URL.qll:
```diff
@@ -0,0 +1,47 @@
+    result = this.getArgument(1)
+  }
+
+  Expr protocolArg() {
```
Missing doc comment

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/URL.qll:
```diff
@@ -0,0 +1,47 @@
+class UrlConstructor extends ClassInstanceExpr {
+  UrlConstructor() { this.getConstructor().getDeclaringType().getQualifiedName() = "java.net.URL" }
+
+  Expr hostArg() {
```
Missing doc comment

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/URI.qll:
```diff
@@ -0,0 +1,43 @@
+  }
+}
+
+class UriCreate extends Call, UriCreation {
```
Missing doc comment

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/URL.qll:
```diff
@@ -0,0 +1,47 @@
+  }
+}
+
+class UrlOpenConnectionMethod extends Method {
```
Missing doc comment

smowton commented on java/ql/src/semmle/code/java/frameworks/javase/URL.qll:
```diff
@@ -0,0 +1,47 @@
+  }
+}
+
+class UrlOpenStreamMethod extends Method {
```
Missing doc comment

smowton commented on java/ql/src/semmle/code/java/frameworks/spring/SpringHttp.qll:
```diff
@@ -38,3 +39,17 @@ class SpringResponseEntityBodyBuilder extends Interface {
+    numArgs = this.getNumArgument()
+  }
+
+  Argument getUriArg() {
```
Missing doc comment

smowton reviewed (commented):
Just noticed some public defns with missing doc comments

aschackmull commented on java/ql/src/Security/CWE/CWE-319/HttpsUrls.ql:
```diff
@@ -29,18 +30,6 @@ class HTTPString extends StringLiteral {
   }
 }
 
-class URLConstructor extends ClassInstanceExpr {
```
When moving this it has been renamed to `UrlConstructor` so presumably something breaks here?

porcupineyhairs replied:
creating a new Pr to address the same.

porcupineyhairs replied:
Sorry, I just checked there are uses of this in `OpenStream.ql` and `InsecureBasicAuth.ql` files but they both define their own `URLConstructor` and use that so we are good here.

aschackmull replied:
So that just means that there are more opportunities for this refactor? In any case, my comment still applies: This code looks broken, if you try to run this query I expect a compilation failure (you've renamed `URL` to `Url`, which I think is fine, but does require some further updates in this file).

porcupineyhairs replied:
@aschackmull The `OpenStream.ql` query is obsoluted by this one. It only include the `OpenStream` call as a sink which I have already included here. There no reason to keep the `OpenStream` PR once this one is merged. As for the `InsecureBasicAuth.ql` file, `@lucha-bc` defines his own set of classes to model URL's. once this Pr is merged, I would open a new PR to remove the redundant code. As for the tests, I can confirm the test suite works fine with the latest set of changes.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/ApacheHttp.qll:
```diff
@@ -13,3 +13,32 @@ class ApacheHttpEntityGetContent extends Method {
+ * Examples include `HttpGet`,`HttpPost` etc.
+ */
+
+class TypeApacheHttpRequest extends Class {
```
This class is almost equal to `TypeApacheHttpRequestBase` - they both include subclasses of `HttpRequestBase` and `BasicHttpRequest`, but this class excludes just those two superclasses. Is there an important reason for this distinction? Otherwise let's just use one class. (And the simpler name `ApacheHttpRequest` seems fine.)

porcupineyhairs replied:
I have directly inlined the class now.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/JaxWS.qll:
```diff
@@ -170,6 +170,13 @@ class JaxRsResponseBuilder extends Class {
 }
 
+/**
+ * The class `javax.ws.rs.client.Client`
```
```suggestion
 * The class `javax.ws.rs.client.Client`.
```

aschackmull commented on java/ql/src/semmle/code/java/frameworks/javase/Http.qll:
```diff
@@ -0,0 +1,20 @@
+import java
+import semmle.code.java.dataflow.FlowSources
```
This import should not be here.
```suggestion
```

aschackmull commented on java/ql/src/semmle/code/java/frameworks/javase/Http.qll:
```diff
@@ -0,0 +1,20 @@
+import java
+import semmle.code.java.dataflow.FlowSources
+
+/** A class representing `HttpRequest.Builder`. */
```
```suggestion
/** The interface `HttpRequest.Builder`. */
```

aschackmull commented on java/ql/src/semmle/code/java/frameworks/javase/Http.qll:
```diff
@@ -0,0 +1,20 @@
+  TypeHttpRequestBuilder() { hasQualifiedName("java.net.http", "HttpRequest$Builder") }
+}
+
+/** A class representing `java.net.http.HttpRequest`. */
```
```suggestion
/** The interface `java.net.http.HttpRequest`. */
```

aschackmull replied:
Not yet updated.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/javase/Http.qll:
```diff
@@ -0,0 +1,20 @@
+  TypeHttpRequest() { hasQualifiedName("java.net.http", "HttpRequest") }
+}
+
+/** A class representing `java.net.http.HttpRequest$Builder`'s `uri` method. */
```
```suggestion
/** The `uri` method on `java.net.http.HttpRequest.Builder`. */
```

aschackmull commented on java/ql/src/semmle/code/java/frameworks/javase/Http.qll:
```diff
@@ -0,0 +1,20 @@
+import java
```
Missing file-level qldoc.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/javase/URI.qll:
```diff
@@ -0,0 +1,43 @@
+import semmle.code.java.dataflow.FlowSources
+
+/** Any expresion or call which returns a new URI.*/
+abstract class UriCreation extends Top {
```
Let's move all the contents of this file to `Networking.qll` - that already contains `TypeUri`.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/javase/URL.qll:
```diff
@@ -0,0 +1,47 @@
+import semmle.code.java.dataflow.FlowSources
+
+/* Am URL constructor expression */
+class UrlConstructor extends ClassInstanceExpr {
```
Let's move all the contents of this file to `Networking.qll` - that already contains `TypeUrl`.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringHttp.qll:
```diff
@@ -38,3 +39,17 @@ class SpringResponseEntityBodyBuilder extends Interface {
+
+/** Models `org.springframework.http.RequestEntity`s instantiation expressions. */
+class SpringRequestEntityInstanceExpr extends ClassInstanceExpr {
+  int numArgs;
```
Remove this field - this is already available as `. getNumArgument()`.

aschackmull replied:
Actually, strike that. Remove this entire class (i.e. inline `SpringRequestEntityInstanceExpr .getUriArg`).

aschackmull commented on java/ql/src/semmle/code/java/frameworks/spring/SpringWebClient.qll:
```diff
@@ -27,3 +27,116 @@ class SpringWebClient extends Interface {
   }
 }
+
+/**
+ * An abstract class representing all Spring Rest Template methods
+ * which take an URL as an argument.
+ */
+abstract class SpringRestTemplateUrlMethods extends Method {
+  /** Gets the argument which corresponds to a URL */
+  abstract Argument getUrlArgument(MethodAccess ma);
+}
```
Lots of issues here. First of, generally avoid public `abstract` classes unless you really want to expose their open-ended semantics. In this case this class and all of its subclasses should be collapsed to just a single class.
Second, is a list of method names the best way to capture this? Or is it better to actually model things as the qldoc describes? I.e. a method on `SpringRestTemplate` with a parameter of type `TypeUri` or `TypeUrl`?
Once we've rewritten the code to that point it makes sense to ask whether this then belongs in the Spring framework library or whether it should be closer to the query (I think the latter is more appropriate if we simply model the methods based on parameter type).

porcupineyhairs replied:
@aschackmull The arguments may or may not be `java.net.URL`'s. I have changed the doc to reflect the same. 
As for the refactoring. I have inlined all classes into a single class and moved it to the `RequestForgeryCustomizations.qll` file. I think it looks much better this way now.

aschackmull reviewed (changes requested):
I have more comments, but let's start with this.

aschackmull commented on java/ql/src/experimental/CWE-918/RequestForgeryCustomizations.qll:
```diff
@@ -0,0 +1,182 @@
+class SpringRestTemplateUrlMethods extends Method {
+  SpringRestTemplateUrlMethods() {
+    this.getDeclaringType() instanceof SpringRestTemplate and
+    this.hasName("doExecute")
+    or
+    this.hasName("postForEntity")
+    or
+    this.hasName("postForLocation")
+    or
+    this.hasName("postForObject")
+    or
+    this.hasName("put")
+    or
+    this.hasName("exchange")
+    or
+    this.hasName("execute")
+    or
+    this.hasName("getForEntity")
+    or
+    this.hasName("getForObject")
+    or
+    this.hasName("patchForObject")
```
```suggestion
    this
        .hasName([
            "doExecute", "postForEntity", "postForLocation", "postForObject", "put", "exchange",
            "execute", "getForEntity", "getForObject", "patchForObject"
          ])
```

aschackmull commented on java/ql/src/experimental/CWE-918/RequestForgeryCustomizations.qll:
```diff
@@ -0,0 +1,182 @@
+  private class ApacheSetUri extends Sink {
+    ApacheSetUri() {
+      exists(MethodAccess ma, TypeApacheHttpRequestBase t |
+        ma.getReceiverType().extendsOrImplements(t) and
```
Why the `.extendsOrImplements`? `TypeApacheHttpRequestBase` already includes subclasses.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/ApacheHttp.qll:
```diff
@@ -13,3 +13,23 @@ class ApacheHttpEntityGetContent extends Method {
+ *  A class derived from the `HttpRequestBase` or the `BasicHttpRequest`
+ *  class of the Apache Http Client `org.apache.http` library
+ */
+class TypeApacheHttpRequestBase extends RefType {
```
Rename to `ApacheHttpRequest`. It includes subclasses, so naming it `Base` is misleading. And the `Type` prefix doesn't add much.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/ApacheHttp.qll:
```diff
@@ -13,3 +13,23 @@ class ApacheHttpEntityGetContent extends Method {
+  }
+}
+
+/* A class representing the `RequestBuilder` class of the Apache Http Client library */
```
Qldoc is missing an additional `*` to make it proper qldoc rather than a simple comment. Also, drop the "A class representing .." style of qldoc - all ql classes represent something, so we might as well state directly what that something is.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/Networking.qll:
```diff
@@ -41,3 +41,88 @@ class SocketGetInputStreamMethod extends Method {
 }
+
+/** Any expresion or call which returns a new URI. */
+abstract class UriCreation extends Top {
```
Avoid public `abstract` classes - they come with a lot of semantic baggage and should be considered an advanced level ql feature not to be used lightly.
Instead you can do
```
class UriCreation extends Call {
  UriCreation() {
    this.getCallee().getDeclaringType() instanceof TypeUri and
    (this instanceof ClassInstanceExpr or this.getCallee().hasName("create"))
  }

  Expr getHostArg() { none() }
}
```
You can then define the two subclasses and override `getHostArg` as you currently do in order to keep that defined via dispatch.

Also, you're not using `UriCreation`? Is that an oversight?

porcupineyhairs replied:
yes that was an oversight.

aschackmull commented on java/ql/src/semmle/code/java/frameworks/Networking.qll:
```diff
@@ -41,3 +41,98 @@ class SocketGetInputStreamMethod extends Method {
+
+/** An URI constructor expression */
+class UriConstructor extends ClassInstanceExpr, UriCreation {
+  UriConstructor() { this.getConstructor().getDeclaringType().getQualifiedName() = "java.net.URI" }
```
This line is now superfluous.
```suggestion
```

aschackmull commented on java/ql/src/semmle/code/java/frameworks/Networking.qll:
```diff
@@ -41,3 +41,98 @@ class SocketGetInputStreamMethod extends Method {
+
+/* An URL constructor expression */
+class UrlConstructor extends ClassInstanceExpr {
+  UrlConstructor() { this.getConstructor().getDeclaringType().getQualifiedName() = "java.net.URL" }
```
Use `hasQualifiedName` instead of `getQualifiedName` here.

aschackmull replied:
Strike that. Use `instanceof TypeUrl` instead.

aschackmull commented:
The qldoc generally needs a lot of updates to bring it in line with https://github.com/github/codeql/blob/main/docs/qldoc-style-guide.md. Note in particular:
* Classes are documented with a noun-phrase, such as `A <thing> that <has properties>.` or `The <some unique thing>.`
* Predicates with results use a third-person verb phrase of the form `Gets (a|the) <thing>.`

porcupineyhairs commented:
@aschackmull I have pushed a few changes. Do I fix them now?

aschackmull commented:
I've written a bunch of additional suggested changes in a PR against this PR. Please merge https://github.com/porcupineyhairs/ql/pull/2 into this PR.

porcupineyhairs commented:
@aschackmull Merged!

Marcono1234 commented:
Is it intended that the files of this query are directly in the `src/experimental` directory instead of `src/experimental/Security/CWE`?

(Sorry if this has been mentioned in one of the review comments; I only went over the pull request comments, but did not find anything regarding this)

smowton commented:
They don't appear to be? What file are you referring to?

Marcono1234 commented:
These:
![Directories screenshot](https://user-images.githubusercontent.com/11685886/107712766-457dd480-6cca-11eb-8c93-35b00bbc8d4a.png)
See https://github.com/github/codeql/tree/main/java/ql/src/experimental

smowton commented:
Doh right, I thought you meant files directly within the `experimental` directory
